Retry callback into job queue / fix simulation_context

xjules opened the pull request:
- fix sumbit test cound and update status
- test queue retries
- test for queue manager and fix to job queue test
- fix for simulation_context and simulation_batch and the tests which use python job now
- integrates and updates #757 

**Issue**
#686 


**Approach**
It includes changing the job status and making sure number of submits is properly set.

ManInFez commented on python/res/job_queue/queue.py:
```diff
@@ -273,6 +278,12 @@ def fetch_next_waiting(self):
+    def count_status(self, status):
+        return len([job for job in self.job_list if job.status == status])
+
+
```
Lots and lots of whitespace..

xjules replied:
gone....

ManInFez commented on python/res/simulator/simulation_context.py:
```diff
@@ -19,43 +15,34 @@ def __init__(self, ert, sim_fs, mask, itr , verbose=False):
+        self._run_context = ErtRunContext( EnkfRunType.ENSEMBLE_EXPERIMENT, sim_fs, None, mask, path_fmt, jobname_fmt, subst_list, itr)
+        for run_arg in self._run_context:
+            if run_arg is not None:
+                self._run_args[run_arg.iens] = run_arg
```
Why are we storing both the run_context and run_args. Ownership of the run_args objects is not very clearly seen here.

xjules replied:
That's for convenience purposes. Agree, this can go away.

ManInFez commented on python/tests/res/simulator/test_batch_sim.py:
```diff
@@ -193,10 +192,8 @@ def test_batch_simulation(self):
             ]
 
             ctx = rsim.start("case", case_data)
+
```
remove whitespace

xjules replied:
ok

ManInFez commented on python/tests/res/simulator/test_simulation_context.py:
```diff
@@ -37,31 +37,20 @@ def test_simulation_context(self):
             while simulation_context2.isRunning():
                 time.sleep(1.0)
 
+
```
remove whitespace

xjules replied:
done

xjules committed:
retry job into queue job

- test queue retries
- test for queue manager and fix to queue test
- fix for job_queue_manager to account for get submits correctly
- remove add_simulation / use of run_simple_step instead
- running sims right form __init__
- fix wait_until in joq_queue
- Update status of non-started jobs in queue_manager
- wait_until in test_stop_sim
- fix test_add_jobs with wait_unitil
- add start_queue in test_add_jobs
- remove _run_arg from simulation_context
- simulation_context need to wait for fs sync
- more descr. in comments and job_status test included
- geo_id back to simulation_context, formatting and cleaning
- add test for geo_id
- fix for JOB_WAITING turns into JOB_KILLED only if stopped by the user
- fix geo_id set before createRunPath

ManInFez committed:
Use queue manager to kill jobs by listening to the queue flag stopped_by_user

Kill job_node process in monitor thread by listedning to stopped_by_user

Update status of non-started jobs in queue_manager

andreabrambilla commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
 
     def stop_queue(self):
-        self._stop_queue( )
+        self.get_job_queue().kill_all_jobs()
```
If you have managed to remove all the calls to the C methods, please remove also the corresponding cwrap declarations at the beginning of the file

xjules replied:
There is a new issue for it: #728

andreabrambilla commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
     def free(self):
         self._free( )
 
+    #resolve ambihuity done vs success
```
typo

xjules replied:
ok

andreabrambilla commented on python/res/simulator/batch_simulator.py:
```diff
@@ -1,5 +1,7 @@
 from ecl.util.util import BoolVector
 
+from threading import Thread
+from time import sleep
```
why are these here?

xjules replied:
Legacy of testing, now it's gone ;)

andreabrambilla commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
+        #Cleanup all not-started jobs
+        for job in job_queue.job_list:
+            if job.status == JobStatusType.JOB_QUEUE_WAITING:
+                job._set_status(JobStatusType.JOB_QUEUE_IS_KILLED)
```
It looks like this PR includes some of the same changes in #757 
I guess you and @ManInFez should fight over it ;)

xjules replied:
That was agreed on standup before @ManInFez went for vacation.

markusdregi commented on python/tests/res/simulator/test_simulation_context.py:
```diff
@@ -37,30 +35,19 @@ def test_simulation_context(self):
-
-            while simulation_context2.isRunning():
-                time.sleep(1.0)
+            time.sleep(3.0)
```
Please remove the `time.sleep` from all the tests and lower the `wait_until` interval to `0.1`. This should not appear in a test. If this is necessary for the code to work we should figure out why.

xjules replied:
~~`time.sleep()` is unfortunately required to sync the `fs` with job states. We could perphaps integrate it into the `wait_until`~~

xjules replied:
This's been fixed now by waiting for the `enkf_simulation_runner.runSimpleStep` to finish.

oyvindeide commented on python/res/job_queue/job_queue_node.py:
```diff
@@ -71,33 +75,40 @@ def is_running(self):
         x.start()
         return x
 
     def stop(self,  driver):
```
`driver` is no longer needed

ManInFez replied:
driver is still used in monitoring, both for updating status and killing the job.

markusdregi replied:
@ManInFez claims that the above comment by himself can be ignored.

xjules replied:
job `stop` is fixed now.

oyvindeide commented on python/res/job_queue/queue.py:
```diff
@@ -89,13 +90,13 @@ def __repr__(self):
         """
         Short doc...
-
+        The @max_submit argument says how many times the jon be submitted (including a failure)
```
typo jon to job

xjules replied:
fixed

oyvindeide commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
         started_job_threads = []
         while job_queue.is_running():
             job = job_queue.fetch_next_waiting()
             while not job_queue.stopped_by_user and job is not None and job_queue.count_running() <= self.max_running():
```
Doesn't this add one job more than it should?:
`job_queue.count_running() <= self.max_running()`, should've been only less than?

markusdregi replied:
I think you are right, however this is outside the scope of the PR. Well spotted though :clap: Make an issue ;)

xjules replied:
yes, as was pointed out by @markusdregi it's outside this scope; mainly we count the run number from `1`

markusdregi commented on python/res/enkf/enkf_simulation_runner.py:
```diff
@@ -91,7 +91,6 @@ def add_job(self, run_arg, res_config, job_queue):
         num_cpu = res_config.queue_config.num_cpu
         if num_cpu == 0:
             num_cpu = res_config.ecl_config.num_cpu
-        
```
Please remove the white space change.

xjules replied:
fixed

markusdregi commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
-        return ts.datetime()
-
-
     def get_job_queue(self):
```
I suggest that we in the future we rewrite this so that `queue` is a private field and then expose a property named `queue` instead of the `get_job_queue`. We could already now, to be compliant with this, use `self.queue` instead of `self.get_job_queue()` internally in job_queue_manager.

xjules replied:
agree, it's fixed now:
```python
@property
def queue(self):
     return self._queue
```

markusdregi commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
     def free(self):
         self._free( )
 
+    #resolve ambiguity done vs success
```
Please turn this comment into a descriptive issue

xjules replied:
Will do ...

markusdregi commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
     def isJobComplete(self, job_index):
-        return self._job_complete( job_index )
+        return not (self.get_job_queue().job_list[job_index].is_running()
+                    or self.get_job_queue().job_list[job_index].status == JobStatusType.JOB_QUEUE_WAITING)
```
Use `self.queue.isJobWaiting(job_index)` instead?

xjules replied:
but then it could be still `running`, i.e., not `waiting` nor `complete` or I might have misunderstood :)

markusdregi commented on python/res/job_queue/job_queue_manager.py:
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
     def max_running(self):
         return self.get_job_queue().get_max_running()
 
+    def max_submit(self):
```
Should we make this `_max_submit`?

xjules replied:
do you mean having it as a property?
```python
@property
def _max_submit(self):
     return return self.queue.get_max_submit()
```

markusdregi commented on python/res/job_queue/job_queue_node.py:
```diff
@@ -27,14 +27,15 @@ class JobQueueNode(BaseCClass):
         self.exit_callback_function = exit_callback_function
         self.callback_arguments = callback_arguments
-    
+        self.stopped_by_user = False
```
Should we perhaps rename this for `stopped`? And in addition, maybe this should be a private member `_stopped`.

xjules replied:
I think it's just a legacy from the past. Agree, this could be renamed.

xjules replied:
fixed now

markusdregi commented on python/res/job_queue/job_queue_node.py:
```diff
@@ -71,33 +75,40 @@ def is_running(self):
                 self.status == JobStatusType.JOB_QUEUE_UNKNOWN) # dont stop monitoring if LSF commands are unavailable
     
-    def job_monitor(self, driver):
+    def job_monitor(self, driver, max_submit):
```
Should we make this private, i.e. `_job_monitor`?

xjules replied:
sure

markusdregi commented on python/res/job_queue/job_queue_node.py:
```diff
@@ -71,33 +75,40 @@ def is_running(self):
         elif self.status == JobStatusType.JOB_QUEUE_WAITING:
             self.started = False
+        elif self.status == JobStatusType.JOB_QUEUE_IS_KILLED:
+            pass
```
Should we add an:
```
else:
    raise AssertionError("Unexpected job status type after running job: {}".format(self.status))

xjules replied:
We can add this, but we should also clear out the job states we want to use

markusdregi commented on python/res/job_queue/queue.py:
```diff
@@ -210,6 +211,9 @@ def get_max_job_duration(self):
     def set_max_job_duration(self, max_duration):
         self._set_max_job_duration(max_duration)
 
+    def get_max_submit(self):
```
Property?

xjules replied:
`max_submit` property now resides in queue only and `queue_manager` pokes queue for it, so `max_submit` function is gone from `queue_manager`

markusdregi commented on python/res/simulator/batch_simulator_context.py:
```diff
@@ -16,7 +16,7 @@ def __init__(self, result_keys, ert, fs, mask, itr):
         super(BatchContext, self).__init__(ert, fs, mask, itr)
         self.result_keys = result_keys
-        self.res_config = ert.resConfig( )
+        self.res_config = ert.resConfig()
```
Please revert white space fixes. This PR is touching enough files as is.

xjules replied:
ok ... just didn't like the emptiness inside :)

markusdregi commented on python/res/simulator/batch_simulator_context.py:
```diff
@@ -78,6 +78,7 @@ def results(self):
         nodes = [ EnkfNode(self.res_config.ensemble_config[key]) for key in self.result_keys ]
         for sim_id in range(len(self)):
             node_id = NodeId( 0, sim_id)
+
```
Please revert white space fixes. This PR is touching enough files as is.

xjules replied:
fixed

markusdregi commented on python/res/simulator/simulation_context.py:
```diff
@@ -91,11 +79,15 @@ def didRealizationFail(self, iens):
 
     def isRealizationQueued(self, iens):
-        return iens in self._run_args
+        try:
```
Are you sure we want to do this?

xjules replied:
we decided not have self._run_args any more (since it's in run_context) - will try to beautify a bit.

markusdregi commented on python/res/simulator/simulation_context.py:
```diff
@@ -123,6 +115,8 @@ def get_run_context(self):
 
     def stop(self):
         self._queue_manager.stop_queue( )
+        if self.sim_thread is not None:
```
Is this necessary? Should not `self.sim_thread` always be a `Thread`?

xjules replied:
ok, that was the legacy from the previous version where `self.sim_thread` was initialized independently. fixed now

markusdregi commented on python/res/simulator/batch_simulator.py:
```diff
@@ -204,9 +204,6 @@ def start(self, case_name, case_data):
         sim_context = BatchContext(
             self.result_keys, self.ert, file_system, mask, itr)
 
-        for sim_id, (geo_id, _) in enumerate(case_data):
```
This scares me a bit. Could you please try to build and run Everest tests on top of this?

markusdregi replied:
In particular, the `geo_id` is lost.

xjules replied:
now `case_data` is an extra argument to `simulation_context`

markusdregi commented on python/res/simulator/simulation_context.py:
```diff
@@ -19,52 +15,44 @@ def __init__(self, ert, sim_fs, mask, itr , verbose=False):
-            raise UserWarning("Realization number: '%d' already queued" % iens)
-
-        run_arg = self._run_context[iens]
-        run_arg.geo_id = geo_id
```
This is crucial and should be preserved. See comment on `geo_id` being lost. Could be passed to the `BatchContext` upon creation.

markusdregi replied:
Note: We should ensure that the tests actually capture this behavior

xjules replied:
Agree, we can sit down and suggest a proper test for capturing the `geo_id`

xjules replied:
Decided just to do a simple `geo_id` assert in `test_simulation_context`

markusdregi commented on python/tests/res/job_queue/test_job_queue.py:
```diff
@@ -1,7 +1,9 @@
 from res.job_queue import JobStatusType, Driver, QueueDriverEnum, JobQueue, JobQueueNode
 from tests import ResTest
+from tests.utils import wait_until
```
Because twice is always better

xjules replied:
agree, twice is always better

markusdregi commented on python/tests/res/job_queue/test_job_queue.py:
```diff
@@ -79,9 +81,13 @@ def test_kill_jobs(self):
-         
-            job_queue.kill_all_jobs()
-            assert not job_queue.is_running()
+
```
Should we add 

```python
time.sleep(3)
wait_until(
    lambda: self.assertTrue(job_queue.is_running())
)
```

to prevent the above `true`  vs `True` bug.

xjules replied:
fixed

markusdregi commented on python/tests/res/job_queue/test_job_queue.py:
```diff
@@ -97,26 +103,43 @@ def test_add_jobs(self):
+            for t in threads:
+                t.join()
+
+            assert True
```
Why?

xjules replied:
fixed now

markusdregi commented on python/tests/res/job_queue/test_job_queue.py:
```diff
@@ -97,26 +103,43 @@ def test_add_jobs(self):
+
+            wait_until(
+                func=(lambda: self.assertFalse(job_queue.is_running())),
+                interval=2
```
Why do we need intervals here?

xjules replied:
Could be removed I reckon.

markusdregi commented on python/tests/res/job_queue/test_job_queue_manager.py:
```diff
@@ -24,16 +25,22 @@ def dummy_exit_callback(args):
 
-def create_queue():
+failing_script = "#!/usr/bin/env python\n"\
+                        "import sys\n"\
```
Should we do something with the line alignment here?

markusdregi replied:
Perhaps this would be more readable?
```python
failing_script = """
#!/usr/bin/env python
import sys
sys.exit(1)
"""
```

xjules replied:
readability fixed

markusdregi commented on python/tests/res/simulator/test_batch_sim.py:
```diff
@@ -405,7 +401,11 @@ def test_stop_sim(self):
 
+            time.sleep(2.0)
             ctx.stop()
+            wait_until(
```
Why is this necessary?

xjules replied:
`context.stop()` is blocking operation now, but needs to be tested a few times first - then it's gone

markusdregi reviewed (commented):
Me and @ManInFez has gone through the entire PR now. Please address the comments.

Well done so far :+1:

ManInFez reviewed (approved):
A very big step in removing C from the queue system. Good work!

markusdregi commented on python/tests/res/simulator/test_simulation_context.py:
```diff
@@ -29,38 +27,28 @@ def test_simulation_context(self):
-            ert.createRunpath( simulation_context1.get_run_context( ) )
-            ert.createRunpath( simulation_context2.get_run_context( ) )
+            # i represents geo_id
+            case_data = [(i,0) for i in range(size)]
```
This is wrong usage of the API. Please revise the documentation: https://github.com/equinor/libres/blob/9ba9839785764568e6479b34005154bfb8b7575f/python/res/simulator/batch_simulator.py#L133

The following should be correct:
```python
case_data = [(i, {}) for i in range(size)]
```

xjules replied:
ok, I knew it was supposed to be a dict, but since it not used anyway ... fixed now
